Update ImGui/Docking

sfulham opened the pull request:
ImGui Update

rokups committed:
Examples: Bring code of example_apple_metal and example_apple_opengl2 closer to each other. (#1873, #3543)

mandar1jn commented:
This really should be done. Because this fixes bugs and issues. Plus there will be a point at which it won't be possible anymore

thedmd committed:
Backends: Replace direct access to TextureId with GetTexID() call in ImDrawCmd. (#3761)

rokups committed:
Backends, Viewports: GLFW: Add a workaround for stuck keys after closing a GLFW window (#3837).

ocornut committed:
Internals: maintaining focus order inside windows + only storing root windows in WindowsFocusOrder[] array. (toward #2304)

ocornut committed:
Viewports, Backends: Vulkan: Rebuild swapchain on VK_SUBOPTIMAL_KHR. (#3881)

ocornut committed:
Scrolling: Fix scroll tracking with e.g. SetScrollHereX/Y() when WindowPadding < ItemSpacing. Fix scroll snapping on edge of scroll region when both scrollbars are enabled.

CalcNextScrollFromScrollTargetAndClamp() fixed snapping edge calculation missing ScrollbarSizes.y

lethal-guitar committed:
Examples: Add OpenGL ES 2.0 support to modern GL examples. (#2837, #3951)

lethal-guitar committed:
Backends: SDL2: Extend global mouse pos availability check (#3950)

AgentX1994 committed:
Backends: Vulkan: Fix mapped memory validation error when buffer sizes are not multiple of VkPhysicalDeviceLimits::nonCoherentAtomSize. (#3957)

If using Vulkan backend and either vertex_size or index_size is not aligned to VkPhysicalDeviceLimits::nonCoherentAtomSize, then the call to vkFlushMappedMemoryRanges in ImGui_ImplVulkan_RenderDrawData will result to validation error.

rokups committed:
TabBar: Use mouse position instead of hardcoded +1/-1 offset when reordering tabs.

Fixes tab reordering in test engine when using fast mode.

rokups committed:
TabBar: Use mouse position instead of hardcoded +1/-1 offset when reordering tabs.

Fixes tab reordering in test engine when using fast mode.

# Conflicts:
#	imgui_widgets.cpp

ocornut committed:
Backends: Win32: Clearing keyboard down array when losing focus (WM_KILLFOCUS). (#2062, #3532, #3961)

ocornut committed:
Examples: Win32: using a more explicit loop for PeekMessage polling to make the code easier to copy and paste and less error-prone.

ocornut committed:
Tables: Expose TableSetColumnEnabled() in public api. (#3935)

ocornut committed:
BeginMainMenuBar(): remove expectation that we don't know menu bar height ahead, allowing up to generalize placement in any direction (will be done in master)

Amend 75de34e281141f851f7e5f9f6a923573167bc0f0

ocornut committed:
Internals: extracted a more reusable BeginViewportSideBar() out of BeginMainMenuBar(). (#3966, #3518)

Complement ca34c81c in docking branch which removed assumption that we can't tell size ahead of Begin().

ocornut committed:
Merge branch 'master' into docking

# Conflicts:
#	backends/imgui_impl_dx9.cpp
#	backends/imgui_impl_sdl.cpp
#	backends/imgui_impl_win32.cpp
#	imgui.cpp
#	imgui_internal.h
#	imgui_widgets.cpp

ocornut committed:
Viewports: Hotfix for crash in monitor array access, caused by 4b9bc4902. (#3967)

ocornut committed:
TabBar: Amend previous commit. Fix tab reordering when tab bar has scrolling.

Some tidying up with helpers + honor 16-bit offsets as with other tab bar features (unlikely single reorder can reach that but consistent)

ocornut committed:
TabBar: Amend previous commit. Fix tab reordering when tab bar has scrolling.

Some tidying up with helpers + honor 16-bit offsets as with other tab bar features (unlikely single reorder can reach that but consistent)

ocornut committed:
ColorEdit4: Alpha default to 255 (instead of 0) when omitted in hex input. (#3973)

Courtesy of @squadack

ocornut committed:
Docking: DockSpace() returns its node ID + adding branch changelog.

utilForever committed:
Fix warning message C4100 (#3974)

- warning C4100: 'fmt_out_size': unreferenced formal parameter

ocornut committed:
Popups: fix comment about OpenPopupOnItemClick(). (#3981)

ocornut committed:
Fix IsWindowAppearing() from returning true twice in most cases. (#3982, #1497, #1061)

ocornut committed:
Docking: removed io.ConfigDockingWithShift option. (#2109)

ocornut committed:
Docking: fix undocking from tab-bar by moving mouse horizontally, broken by d705192.

ocornut committed:
Internals: shallow tidying up of the old resize border structure. Make them match the ImGuiDir order. Split GetWindowResizeID() into GetWindowResizeCornerID() and GetWindowResizeBorderID().

ocornut committed:
Docking: fix undocking from tab-bar by moving mouse horizontally, amend 3ed07a8 + d705192.

Automation system may drag e.g. right-most tab far left (and vice-versa) and one frame and our current logic would fail at it.

ocornut committed:
Internals: storing WindowsHoverPadding in context for external access. + Docs update binaries

ocornut committed:
Added build/ to gitignore (#3027), Examples: SDL2: Link with shell32.lib required by SDL2main.lib since SDL 2.0.12. [#3988, #3884] + added batch file for SDL+Vulkan.

ocornut committed:
Merge branch 'master' into docking

# Conflicts:
#	imgui.cpp
#	imgui_widgets.cpp

ocornut committed:
Fix popup positioning, broken by 84e6fe4. (#3991, #3982)

ocornut committed:
Fix popup positioning, broken by 84e6fe4. (#3991, #3982)

ocornut committed:
Added OpenPopup() ImGuiID overload (#3993, #331)

ptxmac committed:
Update example makefiles to check the new homebrew paths (#4003)

Homebrew on Apple Silicon (i.e. Macbook Air/Pro M1) use `/opt/homebrew` instead of /usr/local.

TotalCaesar659 committed:
Update URLs to HTTPS (#4011)

rokups committed:
Examples: Vulkan: Update GPU selection to pick discrete GPU if available, or use first GPU otherwise. (#4012)

Fixes examples failing on optimus laptops with integrated graphics not supporting Vulkan.

dougbinks committed:
InputText: Do not filter private unicode codepoints (e.g. icons) when pasted from clipboard. (#4005)

s7jones committed:
Backends: OpenGL3: Don't try to read GL_CLIP_ORIGIN unless we're OpenGL 4.5. (#3998, #2366, #2186)

ocornut committed:
Scrolling: Fix mouse wheel axis swap when using SHIFT on macOS (system already does it). (#4010)

ocornut committed:
Fixed clipping of multi-line value text when label is single-line + Fixed vertical alignment of single-line value text when label is multi-line. (#4004)

ocornut committed:
Metrics: hotfix (fix 6ba13349 + f08566b4, #4005)

ocornut committed:
Docking: Fixed restoring of tab order within a dockspace or a split node.

(tests in "docking_tab_order")

ocornut committed:
Docking: Fixed multiple simultaneously reappearing window from appearing undocked in their initial frame.

ocornut committed:
Docking: Fixed reappearing docked windows with no close button showing a tab with extraneous space for one frame.

ocornut committed:
Docking: move NavWindow to SelectedTabId application lower to leave a chance for in-between code to alter focus. + store per-node window menu button id to simplify usage.

ocornut committed:
Docking: Fix window menu button. Broken by 3f16a52 (#4043)

Worked on single-frame click.

ocornut committed:
Documentation tweaks (#4042)

ocornut committed:
Changelog: added docking+entries from 1.72 to 1.82 to increase their visibility.

thedmd committed:
Make PathArcTo accept counter-clockwise angles (#4030, #3491)

duddel committed:
Examples: Android: Make Android build compatible with Gradle 7.0. (#3446)

ocornut committed:
Backends: DirectX10, DirectX11: fixed a crash when backing/restoring state if nothing is bound when entering the rendering function. (#4045)

ocornut committed:
Internals: Added ImFloorSigned() + use our own ImFloorSigned() for stb_truetype, compared as matching output for variety of font data. (#2884)

ocornut committed:
Demo: Improved popups demo and comments.

ocornut committed:
Merge remote-tracking branch 'origin/master' into docking

# Conflicts:
#	backends/imgui_impl_opengl3.cpp

ocornut committed:
Tables: Better preserve column data (mainly widths) when columns count changes. (#4046) + .ini skips columns with no data.

rokups committed:
Backends: OSX: Fix keys remaining stuck in pressed state when CMD-tabbing to a different application. (#3832)

ocornut committed:
ImDrawList: Fixed/improved thickness of thick strokes with sharp angles. (#4053, #3366, #2964, #2868, #2518, #2183)

ocornut committed:
ImDrawList: Revert alteration of normal scaling threshold, for now prioritize preserving property of limiting extents. (#4053, #3366, #2964, #2868, #2518, #2183)

Amend fdda8b8

ocornut committed:
Merge branch 'master' into docking

# Conflicts:
#	imgui.cpp
#	imgui_internal.h

ocornut committed:
Nav: Fixed using SetItemDefaultFocus() on windows with _NavFlattened flag + tools in metrics.

ocornut committed:
Internals: Nav: renaming.

ocornut committed:
Backends: DirectX9: explicitely setting up more graphics states to increase compatibility with unusual non-default states. (#4063)

Added D3DRS_FILLMODE, D3DRS_ZWRITEENABLE, D3DRS_RANGEFOGENABLE, D3DRS_SPECULARENABLE, D3DRS_STENCILENABLE, D3DRS_CLIPPING. Reordered some to match order in DX9 headers.

ocornut committed:
Nav: clear navid in NavInitWindow() when window is marked with ImGuiWindowFlags_NoNavInputs. (#787)

(no expected side-effects, maybe consistency, but those NavLastIds[] values would likely have been 0 already)

rokups committed:
Nav, InputText: Fixed accidental menu toggling while typing non-ascii characters using AltGR. (#370)

ocornut committed:
Nav: clear activeid on menu toggle (so pressing Alt leaving an InputText will steal ActiveID. Previously wouldn't not, but ButtonBehavior() hover logic would not show Menu layer item as hovered when ActiveId != 0). (#787)

(Adjacent to #370)

ocornut committed:
Nav: Fixed pressing Escape to leave menu layer while in a popup or child window. (#787)

Jarhmander committed:
Backends: Win32: Change the case of XInput.h include, so it cross-compile nicely on unices. (#4074)

drglove committed:
Remove const qualifier on ImTextureID local (#4060)

When defining ImTextureID to a type such as Texture*, the const-qualifier here stops us from using the texture ID. const Texture* is not valid to be pushed as a texture ID anymore.

ocornut committed:
CI: test with ImTextureID as pointer to catch accidental reuse of const (#4060)

ocornut committed:
InputText: Align caret/cursor to pixel coordinates. (#4080)

ocornut committed:
Fixes for Visual Studio 2019 static analyzers. (#3938, #4073) + two minor edge case were invalid scalar input (e.g. a sign only) would return buffer as modified.

ocornut committed:
Fixes for PVS Studio and MSVC static analyzers. Using a macro to suppress single-use MSVC false positives. (#3938, #4073)

ocornut committed:
Fixes for MSVC static analyzers (wider range of versions). (#3938, #4073) + warning fix (#4089) + comments.

ocornut committed:
TestEngine: fixed wrong flags passed to ItemInfo queries. Was luckily/misleadingly not causing issues as the test-engine flags we tested were upper bits not colliding with the other type.

Moved those flags higher as well. (unrelated to the fix itself)

ocornut committed:
Internals: ItemFlags is a shared resource as advertised. Fix Begin/PushItemFlags/End sequence (mostly for consistency)

ocornut committed:
Nav: Fixed Tabbing initial activation from skipping the first item if it is tabbable through. (#787)

ocornut committed:
Refactor focusable/tabbalbe item registration toward allowing to handle it prior to clipping (not yet the case) (#343, #4079)

Now performed in ItemAdd(). It can't be trivially moved above clipping effectively because it would require us to scroll to be useful, meaning we'd be better off locking the bounding box a frame earlier. Still wip.
As-is this commit has no value for end-user, but it's a reengineering that moves us closer to the solution. + Internals: moved internal flags.

ocornut committed:
Merge branch 'master' into docking

+ fix warning fix for mingw+dx9
# Conflicts:
#	backends/imgui_impl_dx9.cpp
#	imgui.cpp

ClaudiuHKS committed:
Backends: DX9: Fix potential resource leak (#4093)

ocornut committed:
Disabling some of MSVC most aggressive Debug runtime checks for some simple/low-level functions

(e.g. ImVec2, ImVector) leading to a 10-20% increase of performances with MSVC "default" Debug settings.

wolfpld committed:
Add and use SSE-enabled ImRsqrt() in place of 1.0f / ImSqrt(). (#4091)

Squashed 3 commits.

ITotalJustice committed:
Examples: add backends include path in readmes. (#4106)

mertcandav committed:
Backends: Android: Tweaks. (#4034)

ocornut committed:
Tables: sharing transient buffers between tables, reducing memory footprints. (#3740)

ocornut committed:
Tables: sharing splitter and sort buffers between tables, reducing memory footprints. (#3740)

+ GC pass on that data.

bfierz committed:
Backends: WGPU: update to latest specs. (#4116, #3632)

Merged 13 commits.

ocornut committed:
Tweak computation of io.Framerate so it is less biased toward high-values in the first 120 frames. (#4138)

ocornut committed:
InputText: Fixed CTRL+Arrow or OSX double-click leaking the presence of spaces when ImGuiInputTextFlags_Password is used. (#4155, #4156) [@michael-swan]

ocornut committed:
InputText: Internal renaming to use our own types to clarify code a little bit.

Commit:
Backends: Vulkan: Fix for using IMGUI_IMPL_VULKAN_NO_PROTOTYPES (#4151, #3759, #3227)

ocornut committed:
Backends: WebGPU: Fix build, amend  3c72e51. (#3761)

ocornut committed:
Backends: GLFW: Adding bound check in KeyCallback because GLFW appears to send -1 on some setups. [#4124]

ocornut committed:
Merge branch 'master' into docking

# Conflicts:
#	backends/imgui_impl_dx10.cpp
#	backends/imgui_impl_dx11.cpp
#	backends/imgui_impl_dx12.cpp
#	backends/imgui_impl_dx9.cpp
#	backends/imgui_impl_glfw.cpp
#	backends/imgui_impl_opengl2.cpp
#	backends/imgui_impl_opengl3.cpp
#	imgui.cpp

ocornut committed:
Menus: made MenuItem() in a menu bar reflect the 'selected' argument with a highlight. (#4128)

ocornut committed:
Docs: update links, sponsors

ocornut committed:
Docking: Docking node tab bar honors ItemInnerSpacing.x before first tab. Tweak rendering and alignment of dock node menu marker. (#4130)

+ Fix ~0 in EndFrameDrawDimmedBackgrounds() which is obsolete way of signifying "all round corners".

ocornut committed:
Nav: Fixed single frame CTRL+Tab from properly enabling the  menu layer of target window if it doesn't have other active layers.

ocornut committed:
Docs: Describe IMGUI_DISABLE_FILE_FUNCTIONS in imconfig (#3628, #2734)

ringoz committed:
Backends: WebGPU: Support draw_data->FramebufferScale (#4163)

ocornut committed:
Docs: Improvements to description of using colored glyphes/emojis. (#4169, #3369) + Add Fonts to Metrics. Removed IMGUI_HAS_TABLE markers.

ocornut committed:
Combos: Changed the combo popup to use a different id to also using a context menu with the default item id. (#4167)

ocornut committed:
Fix build with IMGUI_DISABLE_DEMO_WINDOWS

ocornut committed:
Version 1.83

ocornut committed:
Merge branch 'master' into docking

# Conflicts:
#	docs/CHANGELOG.txt

ocornut committed:
Backends: OpenGL3: Handle GL_CLIP_ORIGIN on <4.5 contexts if "GL_ARB_clip_control" extension is detected. (#4170, #3998)

Expecting this to somehow cause another issue but we will fix it when it comes.

ocornut committed:
Metrics, Demo: moved font details display to metrics code. (#4171)

ocornut committed:
Internals: Added ImTextCharToUtf8() helper. Added "out_" markers in more UTF-8 functions. Metrics: Fonts: Fixed display of EllipsisChar by feeding proper utf-8.

ocornut committed:
Version 1.84 WIP

rokups committed:
CI: Add a build of imgui_impl_vulkan.cpp with IMGUI_IMPL_VULKAN_NO_PROTOTYPES defined. (#4151, #3759, #3227)

ocornut committed:
Fix warnings with VS2019 in C++20 mode (#4173)

ocornut committed:
Merge branch 'master' into docking

# Conflicts:
#	backends/imgui_impl_opengl3.cpp
#	docs/CHANGELOG.txt

ocornut committed:
TabBar: Fixed using more than 32 KB-worth of tab names. (#4176)

ocornut committed:
Backends: Comments.

ocornut committed:
Nav, Combo: removed unnecessary window->LastNavIds[] assignment in combo code. + minor renaming.

ocornut committed:
Internals: Combo: Extracted into a BeginComboPopup() function. (#4168)

Renamed frame_bb > bb locally.

ocornut committed:
Fixed printf-style format checks on non-MinGW flavors. (#4183, #3592)

ocornut committed:
Docking: Fix IsWindowAppearing() and ImGuiCond_Appearing on docked windows. (#4177, #3982, #1497, #1061)

ocornut committed:
Docking: Amend 91704b7, window->DockXXX booleans not properly cleared when window not docked. (#4177, #3982, #1497, #1061)

Fix issue with freshly split windows/nodes incorrectly returning true to IsWindowAppearing().

ocornut committed:
Docking: Fix IsWindowAppearing() unnecessarily returning true twice in a row. (#4177, #3982, #1497, #1061) + added a zealous assert.

ocornut committed:
Docking: Clicking on the right-most close button of a docking node closes all windows. (#4186)

ocornut committed:
Internals: fixed iterating ImPool, fix after use of TableRemove() (was only used by TestEngine)

Merge metrics bits from 646c87359

ocornut committed:
Docking: comments (#4189)

ocornut committed:
Internals: fixed ImPool:: iteration, rename GetBufSize to GetMapSize. Amend c6c82b9f

ocornut committed:
Merge branch 'master' into docking

# Conflicts:
#	imgui.cpp
#	imgui_internal.h
#	imgui_widgets.cpp

tatsy committed:
Backends: OpenGL3: Fix access violation due to NULL from glGetStringi. (#4201)

ocornut committed:
Tables: Added ImGuiTableColumnFlags_NoHeaderLabel to request TableHeadersRow() to not submit label for a column. (#4206)

ocornut committed:
Tables: made TableUpdateBorders() use IsVisibleX flag. comments.

ocornut committed:
Tables: Fix columns order on TableSetupScrollFreeze() if previous data got frozen columns out of their section.

ocornut committed:
Tables: offset and shuffle flags (breaks ABI compatibility as often)

If you rely on ABI compatibility consider reworking how your backends are created, using cimgui metadata output. We will happily often break ABI compat.

ocornut committed:
Tables: Added ImGuiTableColumnFlags_Disabled acting a master disable over (hidden from user/context menu). (#3935, #3740)

ocornut committed:
Tables: Clarified that TableSetColumnEnabled() requires the table to use the ImGuiTableFlags_Hideable flag. (#3935)

FunMiles committed:
Backends, Viewports: Vulkan: Fix the use of the incorrect fence in wait for fence. (#4208)

The fence being waited upon was not the one associated with the current frame.
This results in validation error detecting a reset of command buffers still in use and resetting fences while still in use.
Read more details in #4208

ocornut committed:
Examples: update all VS project files to VS2015. Update DX12 project Windows SDK to latest.

ocornut committed:
Examples: SDL2: Accomodate for vcpkg install having headers in SDL2/SDL.h vs SDL.h + vcpkg related comments.

ocornut committed:
Backends: Win32: Rework to handle certains Windows 8.1/10 features without a manifest. (#4200, #4191)

ocornut committed:
Merge branch 'master' into docking

# Conflicts:
#	backends/imgui_impl_win32.cpp

ocornut committed:
Added PushDisabled(), PopDisabled() currently only exposed in imgui_internal.h (#211)

ocornut committed:
Added PushDisabled(), PopDisabled() currently only exposed in imgui_internal.h (#211)

ocornut committed:
ImVector: added clear_delete(), clear_destruct() helpers.

# Conflicts:
#	imgui.cpp

ocornut committed:
ImVector: added clear_delete(), clear_destruct() helpers.

ocornut committed:
Nav, Drag and Drop, Docking: fixed two issues leading nav result to conflict with moving a window. (#4211, #3025)

ocornut committed:
Backends: DX12: Fix texture casting crash on 32-bit systems (introduced on 2021/05/19 and v1.83) + added comments about building on 32-bit systems. (#4225)

ocornut committed:
Metrics: Tentative fix for bad printf format.

Ref b53b8f58dfbd4cb9f9b2a62e7a0a7e4500c491e9, a7a1b3b0a78f64e704c8e3a8b56c6c17bc19b306

ocornut committed:
Internals/experimental: BeginComboPreview(), EndComboPreview(). (#4168, #1658)

(amended)

ocornut committed:
Minor optimization, removing do/while(0) patterns in some cases as they translate to runtime

(e.g. translate to xor + test + jne in VS x84, unnecessary)

rokups committed:
SplitterBehavior: fix using IsItemHovered() after SplitterBehavior()

ocornut committed:
Tables: Fix invalid data in TableGetSortSpecs() when SpecsDirty flag is unset. (#4233)

Amend 4ce6bd8cf, but with usage of ImPool<> bug existed even before 4ce6bd8c. Would only materialize if user called (ableGetSortSpecs and used data without checking SpecsDirty.

ocornut committed:
Viewport: extracted code out of Begin() into WindowSyncOwnedViewport() - no other change

ocornut committed:
Viewports: Fix popup/tooltip created without a parent window from being given a ParentViewportId value of the implicit/fallback window. (#4236, #2409)

Amend 3ead9820

sugrob9000 committed:
Backends: OpenGL3: Destroy shader objects right away (#4244)

In OpenGL, after a shader program has been linked, shader objects don't need to be kept neither attached to the program nor in existence

ocornut committed:
Added IMGUI_DISABLE_SSE (#4250, #4091)

ocornut committed:
Nav: moved RenderNavHighlight() calls of TreeNode and Selectable out of if (hovered || selected) tests. Should make no difference as NavId currently returns hovered. (#1861, #4242)

ocornut committed:
Default window focus scope not 0. Added ImGuiSelectableFlags_SelectOnNav with comments and caveats. (#1861, #4242,)

Focus scope default value: amend 7ee623d9 a5041c88 2ebe08be

rokups committed:
Backends: OSX: Added a fix for shortcuts using CTRL key instead of CMD key. (#4253)

ocornut committed:
ImGuiWindowFlags_UnsavedDocument/ImGuiTabItmeFlags_UnsavedDocument display a dot instead of a '*'.

ocornut committed:
FIx static analysers warnings and disable false positives.

ocornut committed:
Fonts: Functions with a 'float size_pixels' parameter can accept zero if it is set in ImFontSize::SizePixels.

ocornut committed:
ImFontAtlas: move implicit AddFontDefault call to Build() function + remove unnecessary asserts in backend.

ocornut committed:
Merge branch 'master' into docking

# Conflicts:
#	backends/imgui_impl_opengl3.cpp
#	imgui.cpp
#	imgui.h
#	imgui_demo.cpp

harry75369 committed:
Backends: OpenGL3: Enable OES_vertex_array_object extension for emscripten/wasm build (#4266, #4267)

rokups committed:
Examples: example_apple_opengl2: Fix keys remaining stuck in pressed state when using shortcuts with CMD. (#4253)

Fix follows event capture scheme of example_apple_metal, where this issue is not present.

NZJenkins committed:
Backends: DX9: explicitely disable texture state stages after >= 1. (#4268)
imgui only uses texture stage state for stage 0
But any state set in stages >= 1 will affect rendering
Set stage 1 COLOROP to DISABLED, to deactivate stages >= 1

ocornut committed:
Fonts: Prefer using U+FFFD character for fallback instead of '?'. Use U+FF0E dot character to construct an ellipsis if U+002E '.' is not available. (#4269)

ocornut committed:
Fonts: Add U+FFFD ("replacement character") to default asian glyphs ranges. Remove SetFallbackChar() (amend 0ffd99d31). (#4269)

ocornut committed:
Comments + adding spacing in headers because VS IDE parser display blocks so badly. Add helper in internal's ImGuiInputTextState. (#4275)

ocornut committed:
Backends: moved global to a data structure to facilitate support for multi-contexts. (#586, #1851, #2004, #3012, #3934, #4141)

This is NOT enable multi-contexts for any backends
- in order to make this commit as harmless as possible, while containing all the cruft/renaming
-

ocornut committed:
Backends: amends to 1db1066 + merge minor bits from docking incl SetActiveIdUsingNavAndKeys().

No need to clear fields before deletion. DX12: renamed to match docking branch.

ocornut committed:
Backends: amends to 1db1066 + merge minor bits from docking incl SetActiveIdUsingNavAndKeys().

No need to clear fields before deletion. DX12: renamed to match docking branch.

ocornut committed:
Merge branch 'features/backends_context' into docking

# Conflicts:
#	backends/imgui_impl_dx10.cpp
#	backends/imgui_impl_dx11.cpp
#	backends/imgui_impl_dx12.cpp
#	backends/imgui_impl_dx9.cpp
#	backends/imgui_impl_glfw.cpp
#	backends/imgui_impl_opengl2.cpp
#	backends/imgui_impl_opengl3.cpp
#	backends/imgui_impl_sdl.cpp
#	backends/imgui_impl_vulkan.cpp
#	backends/imgui_impl_win32.cpp
#	examples/example_apple_opengl2/main.mm

ocornut committed:
Backends: Viewports: renamed viewport storage structures ImGuiViewportDataXXXX -> ImGui_ImplXXXX_ViewportData and locals (matching naming convention in 70c60385)

ocornut committed:
Backends: support for multiple imgui context with little testing (#586, #1851, #2004, #3012, #3934, #4141)

I believe more renderer backends should work. GLFW/Win32/SDL/Vulkan probably have many issues.

ocornut committed:
Fonts: Fix calling ClearTexData() (clearing CPU side font data) triggering an assert in NewFrame(). (#3487) + Backends: added additional assert to facilitate detecting user understand they haven't initialized a backend.

funchal committed:
Backends: SDL2: Remove unnecessary ImGui_ImplSDL2_NewFrame() parameter. (#3244)

ocornut committed:
Backends: Cleanup, removed unnecessary create/destroy wrappers. Fix allegro5 backend + use same code as other backend.

+ Update gallery links (#4280)

ocornut committed:
Drag and Drop: drop target highlight doesn't try to bypass host clipping rectangle. (#4281, #3272)

Revert 4b94738c

ocornut committed:
Docking: removed DockNodeFlagsOverrideClear flags from ImGuiWindowClass. (#2999, #3521, #3633)

+ extraded bits of metrics into DebugNodeDockNodeFlags()

ocornut committed:
Docking: Reworked node flags saving/inheritance... (#4292, #3834, #3633, #3521, #3492, #3335, #2999, #2648)

..so that flags enforced by docked windows via the DockNodeFlagsOverrideSet mechanism are are not left in empty dockspace nodes once the windows gets undocked.

ocornut committed:
Docking: Added ImGuiDockNodeFlags_NoDockingOverEmpty. Breaking definition of ImGuiDockNodeFlags_NoDockingOverOther which now means "non empty node". (#3492, #2648, #4292)

ocornut committed:
Fix warning + apply trailing whitespaces in stb_ libs (as applied on nothings/stb on 2020/02/02, facilitating further diffs)

ocornut committed:
InputText: Fix handling of paste failure (buffer full) which in some cases could corrupt the undo stack. (#4038)

ocornut committed:
Backends: GLFW: Installing and exposed ImGui_ImplGlfw_MonitorCallback() for forward compatibility with docking branch.

+ Comments (#3934)

ocornut committed:
Docking: Fixed crash issues using DockBuilderRemoveNode() in some situations. (#3111, #3179, #3203, #4295)

If the deleted root node isn't part of a dockspace with a central node, it won't be "protected" but removed when last window gets removed.

ocornut committed:
Docking: Fix crash when a dock node gets re-qualified as dockspace>floating>dockspace.. (#3203, #4295)

Which tends to happen when incorrectly calling DockBuilderAddNode() without ImGuiDockNodeFlags_Dockspace and using it as a Dockspace on the next frame after the floating window hosting the node has been automatically created.

ocornut committed:
Internals: Menus: minor tidying up + renaming in ImGuiMenuColumns + removing extraneous offset field which is always zero + using smaller types.

sizeof() 36 -> 20

ocornut committed:
Internals: Menus: Naming offset field + fix spacing/offset computation to handle more offsets.

ocornut committed:
Backends: Win32: IME functions are disabled by default for non-Visual Studio compilers (MinGW etc.) (#2590, #738, #4185, #4301)

Enable with '#define IMGUI_ENABLE_WIN32_DEFAULT_IME_FUNCTIONS' for those compilers. Undo change from 1.82.

ocornut committed:
Menus: rework to allow for an icon column (not yet exposed, but usable via internals) + fix menus being affected by style.SelectableTextAlign (#126)

ocornut committed:
Update FAQ

ocornut committed:
PushDisabled(): added bool + clarify support for stacked disabled (#211)

ocornut committed:
Nav: Disabled items are not candidate for default focus. (#211, #787) + simplify handling of ImGuiButtonFlags_PressedOnDragDropHold path.

ocornut committed:
BeginMenu: remove unnecessary call to ItemHoverable() which is already done by the Selectable() call.

Amend 88d7b3de

ocornut committed:
InputInt/InputFloat: When used with Steps values and _ReadOnly flag, the step button look disabled. (#211)

ocornut committed:
Internals: Removed ImGuiButtonFlags_Disabled (which had inconsistent behavior) in favor of ImGuiItemFlags_Disabled. Selectable()'s ImGuiSelectableFlags_Disabled now uses the later. (#211)

ocornut committed:
Disabled items more consistently release active id if the active item got disabled. (#211)

ocornut committed:
Disabled: don't prevent Selectable() from being selected. (#211) + BeginMenu()/MenuItemEx() uses PushDisabled() the standard way.

ImGuiSelectableFlags_Disabled is now unused by the library and might be marked obsolete later.

rokups committed:
Disabled: disabled items set HoveredId, allowing e.g. HoveredIdTimer to function. (#211, #3419) + Menus: fix hovering a disabled menu or menu item not closing other menus.

Rework of https://github.com/rokups/imgui/commit/c24b470
Note that the declared intent of that commit "Prevents window from being dragged if mouse hovers a disabled item." was already fullfilled by a876ad87.
Changes in ButtonBehavior() not needed anymore since ImGuiButtonFlags_Disabled is gone

ocornut committed:
Disabled: fixed IsItemHovered() if popped disabled state after item, or when using Selectable_Disabled. (#211)

ocornut committed:
Merge branch 'master' into docking

# Conflicts:
#	backends/imgui_impl_dx10.cpp
#	backends/imgui_impl_dx11.cpp
#	backends/imgui_impl_dx12.cpp
#	backends/imgui_impl_dx9.cpp
#	backends/imgui_impl_glfw.cpp
#	backends/imgui_impl_opengl3.cpp
#	backends/imgui_impl_sdl.cpp
#	backends/imgui_impl_vulkan.cpp
#	backends/imgui_impl_win32.cpp
#	imgui.cpp
#	imgui_internal.h

PathogenDavid committed:
Added missing IMGUI_API to GetViewportPlatformMonitor. (#4309)

ocornut committed:
Disabled: fixed IsItemHovered() returning true on disabled item when navigated to. (#211)

ocornut committed:
Disabled: items more consistently release active id if the active item got disabled. (#211) (amend 2952525)

ocornut committed:
TestEngine, MenuItem: call IMGUI_TEST_ENGINE_ITEM_INFO() at the right stack level. (amend f8fae02)

rokups committed:
Popups: Fix OpenPopup() being called after BeginPopup() resulting in a failure to open a popup when focus is NULL. (#4308)

Story of removed line begins in commit b80cf0a. It's purpose was to close popups when they lost focus.
Later on few other changes were introduced:
* bcc49ff - closing popups with RMB without changing window focus
* af679a1 - closing popups in FocusWindow()
These two changes covered all the cases of popup closing which made deleted line obsolete. Conveniently, now all popup closing is handled from within UpdateMouseMovingWindowEndFrame() either by calling FocusWindow() or ClosePopupsOverWindow().

ocornut committed:
Internals: ItemAdd: set LastItemXXX fields before navigation calls + comments about io.IniFilename (#4294)

ocornut committed:
Expose BeginDisabled()/EndDisabled() in public API. Add to demo. (#211)

ocornut committed:
Internals: moved LastItem data to a shared structure (instead of one per window)

(should be a no-op as we are restoring things in Begin/End. Toward faciliate backup/restore of LastItemData and favor pulling from here instead of CurrentItemFlags, toward #211)

ocornut committed:
Internals: widgets always read back from g.LastItemData.InFlags (so we can now modify per-item disable state more easily). (#211)

ocornut committed:
Fix - amend 1ad1530 (range-select branch need the NextItemData in NavProcessItem)

ocornut committed:
Merge branch 'master' into docking

# Conflicts:
#	imgui.cpp
#	imgui_internal.h

PathogenDavid committed:
Examples: DirectX12: Fixed Alt+Enter fullscreen in DirectX12 example. (#4346, #4348)

This also removes unnecessary recreation of backend-owned device objects when the window is resized.
+ amend original PR with a g_pSwapChain->SetFullscreenState(false, NULL); call.

sitic committed:
Correct version 1.83 release date (#4355)

ocornut committed:
Removed unnecessary line in Begin() + commented out redirecting functions/enums names that were marked obsolete in 1.67 and 1.69 + readded commented obsoleted function.

ocornut committed:
Internals: ColorEdit: renamed some internal masks to match our common coding style.

ImGuiColorEditFlags__DisplayMask > ImGuiColorEditFlags_DisplayMask_, ImGuiColorEditFlags__OptionsDefault > ImGuiColorEditFlags_DefaultOptions_, ImGuiColorEditFlags__DataTypeMask > ImGuiColorEditFlags_DataTypeMask_  etc.

ocornut committed:
Allow imgui_internal.h to be included before imgui.h

(practically speaking it allows me to use StructLayout without hassle)

ocornut committed:
Update FAQ to clarify conflict with window (#4158, #4008, #2562)

ocornut committed:
Backends: SDL2: Mouse position is correctly reported when the host platform window is hovered but not focused. (#2445, #3751, #4377)

(requires SDL 2.0.5 because SDL_GetMouseFocus() is only usable with SDL_HINT_MOUSE_FOCUS_CLICKTHROUGH).

ocornut committed:
Backends: GLFW: Mouse position is correctly reported when the host platform window is hovered but not focused. (#3751, #4377, #2445)

ocornut committed:
Backends: Win32: Mouse position is correctly reported when the host platform window is hovered but not focused. (#2445, #2696, #3751, #4377)

ocornut committed:
Merge branch 'master' into docking

# Conflicts:
#	backends/imgui_impl_glfw.cpp
#	backends/imgui_impl_sdl.cpp
#	backends/imgui_impl_win32.cpp
#	imgui.cpp

ocornut committed:
Backends: Win32, SDL, GLFW: only honor io.WantSetMousePos when focused + fix GLFW uninstalling handler + tweaks to reduce branch drift with docking. (#787, #2445, #2696, #3751, #4377)

ocornut committed:
Backends: Win32, SDL, GLFW: only honor io.WantSetMousePos when focused + fix GLFW uninstalling handler + tweaks to reduce branch drift with docking. (#787, #2445, #2696, #3751, #4377)

# Conflicts:
#	backends/imgui_impl_glfw.cpp
#	backends/imgui_impl_sdl.cpp
#	backends/imgui_impl_win32.cpp

ocornut committed:
Backends: Win32: Fixed keyboard modifiers being reported when host window doesn't have focus. (#2622)

thedmd committed:
IO, Backends: add io.AddFocusEvent(). Clear pressed keys after loosing input focus (#3532)

Amend/fix #2445, #2696, #3751, #4377

kfsone committed:
InputText: Fix named filtering flags disabling newline or tabs in multiline inputs (#4409, #4410)

ocornut committed:
Reword file list as *.cpp *.h (#4413)

thedmd committed:
ImDrawList: Fix AddCircle{Filled} with 12 segment case (#4419, #4421)

ocornut committed:
Nav: Alt doesn't toggle menu layer if other modifiers are held. (#4439)

ocornut committed:
Nav: Fix not clearing NavWindowingToggleLayer properly (old code left it to true, relied on Alt release only). Removed unnecessary KeyMenu_ from NavInput. (#4439, #787)

JJCUBER committed:
Fixed grammar mistake in FAQ (#4431)

ocornut committed:
Backends: OpenGL3: Clarify use of GL_POLYGON_MODE.

rokups committed:
Metrics: Fixed a crash when inspecting the individual draw command of a foreground drawlist.

ocornut committed:
IO: set io.KeyModsPrev in EndFrame() to tolerate backend writing to io.KeyMods without breaking. (#4439)

Amend ff428f1

ocornut committed:
Fix tabbing in/out of an InputTextMultiline(), broken by 6b8a059.

ocornut committed:
Fix various warnings (#4442)

ocornut committed:
Drag and Drop: fix using AcceptDragDropPayload() with ImGuiDragDropFlags_AcceptNoPreviewTooltip.

Window was not properly hidden in that case.

rokups committed:
Backends: OpenGL3: Embed our own minimal GL loader based on gl3w and use it in SDL/GLFW examples. (#4445)

Loader repository at https://github.com/dearimgui/gl3w_stripped
Removed support for using custom loaders in examples/opengl3 backend

ocornut committed:
Backends: OpenGL3: Embed our own minimal GL loader (amends). (#4445)

ocornut committed:
Merge branch 'master' into docking (will need further for io.AddFocusEvent)

# Conflicts:
#	backends/imgui_impl_glfw.cpp
#	backends/imgui_impl_opengl3.cpp
#	backends/imgui_impl_sdl.cpp
#	backends/imgui_impl_win32.cpp
#	imgui.cpp

ocornut committed:
IO: modify io.AddFocusEvent() to tolerate in/out for multi-viewports. Amend 2f40be6. (#3532)

rokups committed:
Backends: OpenGL3: Update loader to include APIs used by test framework.

ocornut committed:
Small optimizations to BeginDisabled() to allow frequent calls (#211)

Not intended for frequent calls but I suspect some people will do it either way...
Rough/indicative: measured 0.1 ms for 5000 calls in release, 0.5 ms in debug on my desktop windows.

ocornut committed:
Version 1.84

Fix PVS Studio false positive //-V1020
Fix missing #ifndef for IMGUI_IMPL_OPENGL_LOADER_CUSTOM path

ocornut committed:
Merge branch 'master' into docking

# Conflicts:
#	docs/CHANGELOG.txt

ocornut committed:
Fix BeginDisabled(false), (#211, #4452)

ocornut committed:
Fix BeginDisabled(false), again, (#211, #4452, #4453) Version 1.84.1

(forced pushed since our earlier versioning didn't sort correctly in github web)

# Conflicts:
#	docs/CHANGELOG.txt

ocornut committed:
Backends: GLFW: Fixed unused variable warning for empty assert macro. (#4459)

ocornut committed:
Docking: fixed settings load issue when mouse wheeling. (#4310)

ocornut committed:
Docking: fix 58f5092 (#4310)

If we clear _ChildWindow flag we must remove it from here otherwise render loop will fail.
